framework/src/play/PropertiesConfLoader.java

```diff
@@ -33,9 +30,33 @@ public Properties readOneConfigurationFile(String playId, String filename) {
   }
 
   private Properties readOneConfigurationFile(String filename, String playId, String inheritedId, Set<VirtualFile> confs) {
-    VirtualFile conf = VirtualFile.open(Play.applicationPath + "/conf/" + filename);
+    VirtualFile conf = resolveFilenameToVirtualFile(filename);
+    return readOneConfigurationFile(conf, playId, inheritedId, confs);
+  }
+
+  protected VirtualFile resolveFilenameToVirtualFile(String filename) {
+    VirtualFile conf = resolveFileNameWithApplicationPath(filename);
+    if (!conf.exists()) {
+      try {
+        conf = resolveFileNameWithClasspath(filename);
+      } catch (NullPointerException e) {
+        // keep going! VirtualFile's exception is more descriptive
+      }
+    }
+    return conf;
+  }
+
+  protected final VirtualFile resolveFileNameWithApplicationPath(String filename) {
+    return VirtualFile.open(Play.applicationPath + "/conf/" + filename);
+  }
+
+  protected final VirtualFile resolveFileNameWithClasspath(String filename) {
+    return VirtualFile.open(Objects.requireNonNull(Thread.currentThread().getContextClassLoader().getResource("conf/" + filename)).getPath());
```

Comment:
`VirtualFile` needs more love: it doesn't knows jars. :confused: 

```java
Exception in thread "main" play.exceptions.UnexpectedException: Failed to read /file:/app/classpath/app.jar!/conf/application.conf
	at play.vfs.VirtualFile.inputstream(VirtualFile.java:106)
	at play.libs.IO.readUtf8Properties(IO.java:17)
	at play.PropertiesConfLoader.readOneConfigurationFile(PropertiesConfLoader.java:63)
	at play.PropertiesConfLoader.readOneConfigurationFile(PropertiesConfLoader.java:34)
	at play.PropertiesConfLoader.readOneConfigurationFile(PropertiesConfLoader.java:29)
	at play.PropertiesConfLoader.readConfiguration(PropertiesConfLoader.java:25)
	at play.Play.readConfiguration(Play.java:260)
	at play.Play.init(Play.java:149)
```

Reply:
@xabolcs Yes, initially in Play! framework `VirtualFile` intention was to resolve files from the project itself and its `modules`. It was not intended to find files in classpath.

Reply:
@asoIntsev, thanks for replying.

Yeah, these exceptions suggest that too.
How could we continue? For me it would be the best that the Framework could work with `jar`s and classpath also ... without breaking `VirtualFile`'s Play! 1 compatibility.
Of course the easiest way is to document these limitations for the RePlay App developers.
Of course, there are solutions in between too, like the current state! 🙂


Yeah, as in the `log4j.properties` PR, I use and pack my App's `conf` folder as a resource folder (becasuse my Idea hides a few folder if "they are too overlapping")

Reply:
@xabolcs I started reviewing the PR. Sorry for a long delay.

I have lots of doubts at the moment... 
1. When you call `openWithEntryName`, you always pass the same name twice its parameters. Duplication.
2. this change actually breaks `VirtualFile` API in that sense that many its methods will not work anymore.

```java
  @Test
  void broken() {
    File file = new File(Thread.currentThread().getContextClassLoader().getResource("org/opentest4j/AssertionFailedError.class").getFile());
    VirtualFile virtualFile = VirtualFile.openWithEntryName(file, "org/opentest4j/AssertionFailedError.class");

    virtualFile.child("some.txt"); // ?
    virtualFile.length(); // 0
    virtualFile.hashCode(); // ?
    virtualFile.lastModified(); // 0
    virtualFile.exists();  // FALSE!
    virtualFile.list(); // null ?
  }
```

Reply:
Yup, the `VirtualFile` part is just a trial and error hack! See my previous comments.

Reply:
Anyway, I'll try to incorporate this failing test and try to not break it with the first commit. :slightly_smiling_face:

Reply:
@xabolcs The more I think about it, the more I come to a conclusion that `VirtualFile` is an **alternative** for classpath. Either you find a file from classpath, or from `VirtualFile`. 

So my suggestion is: 
1. It's generally a good idea to find resources from classpath, but
2. Let's try to find them from two sources: classpath and `VirtualFile`. 
  2.1. I mean, method for searching in classpath should be outside of `VirtualFile`.

Reply:
Yeah, and we are back to my first [comment](https://github.com/codeborne/replay/pull/202#issuecomment-1608561939):
> About the solution: as you can see the pattern is the following. Those startup files are resolved against `Play.applicationPath` which defined as the current dir (`System.getProperty("user.dir")`) and if doesn't exists I try to resolve them against Classpath.
If they not found on the Classpath I leave the workflow intact - the `VirtualFile`'s `UnexpectedException` is more descriptive.
> 
> Could we improve something about this pattern? How about extending `VirtualFile`? One resolves the files against `Play.applicationPath` the other uses Classpath. 🙂


As I'm just a Java copy-paster and not a developer, I don't have solid thoughts how to achieve this "try to resolve the file first by `VirtualFile` then search it on the Classpath" think.
Extracting `VirtualFile`'s methods of read-only functionality to an interface and create an `AbstractFileResolver` which resolves to `VirtualFile` or `ClasspathFile`? :grinning:
Sure there are other and better solution to this.

Reply:
Yes, I was thinking something similar.
But maybe it's even more simpler. Currently you only need to find one file (e.g. `conf/routes`) - either from classpath or from VirtualFile.

Maybe we could just add a static method to `VirtualFile`? Say, `find`:
```java
public static URL routes;
...
routes = appRoot.find("conf/routes");
```

```java
class VirtualFile {
  public static URL find(String name) {
    URL result = // find from classpath
    if (result == null) {
      result = // find from VirtualFile as previously
    }
    return result;
  }
}
```

Reply:
Cool! I'll go that way and see what happens!
Thanks for the hint!